launcherAPI.js

```diff
@@ -0,0 +1,214 @@
+    },
+
+    _acquireUnityDBus: function () {
+        if (!this._unity_bus_id) {
+            Gio.DBus.session.own_name('com.canonical.Unity',
+                Gio.BusNameOwnerFlags.ALLOW_REPLACEMENT, null, null);
+        }
+    },
+
+    _releaseUnityDBus: function () {
+        if (this._unity_bus_id) {
+            Gio.DBus.session.unown_name(this._unity_bus_id);
+            this._unity_bus_id = 0;
+        }
+    },
+
+    _onEntrySignalReceived: function (connection, sender_name, object_path,
+        interface_name, signal_name, parameters, user_data) {
+        if (!parameters || !signal_name)
+            return;
+
+        if (signal_name == 'Update') {
+            if (!sender_name) {
+                return;
+            }
+
+            this._handleUpdateRequest(sender_name, parameters);
+        }
+    },
+
+    _onDBusNameOwnerChanged: function (connection, sender_name, object_path,
+        interface_name, signal_name, parameters, user_data) {
+        if (!parameters || !this.size())
+            return;
+
+        let [name, before, after] = parameters.deep_unpack();
+
+        if (!after) {
+            if (this._entriesByDBusName.hasOwnProperty(before)) {
+                this.removeEntry(this._entriesByDBusName[before]);
+            }
+        }
+    },
+
+    _handleUpdateRequest: function (senderName, parameters) {
+        if (!senderName || !parameters) {
+            return;
+        }
+
+        let [appUri, properties] = parameters.deep_unpack();
+        let appId = appUri.replace(/(^\w+:|^)\/\//, '');
+        let entry = this.lookupByDBusName(senderName);
+
+        if (entry) {
+            entry.setDBusName(senderName);
+            entry.update(properties);
+        } else {
+            let entry = new LauncherEntryRemote(senderName, appId, properties);
+            this.addEntry(entry);
+        }
+    },
+});
+
+Signals.addSignalMethods(LauncherEntryRemoteModel.prototype);
+
+var LauncherEntryRemote = new Lang.Class({
+    Name: 'DashToDock.LauncherEntryRemote',
+
+    _init: function (dbusName, appId, properties) {
+        this._dbusName = dbusName;
+        this._appId = appId;
+        this._count = 0;
+        this._countVisible = false;
+        this.update(properties);
+    },
+
+    appId: function () {
+        return this._appId;
+    },
+
+    dbusName: function () {
+        return this._dbusName;
+    },
+
+    count: function () {
+        return this._count;
+    },
+
+    setCount: function (count) {
+        if (this._count != count) {
+            this._count = count;
+            this.emit('count-changed', this._count);
+        }
+    },
+
+    countVisible: function () {
+        return this._countVisible;
+    },
+
+    setCountVisible: function (countVisible) {
+        if (this._countVisible != countVisible) {
+            this._countVisible = countVisible;
+            this.emit('count-visible-changed', this._countVisible);
+        }
+    },
+
+    setDBusName(dbusName) {
```

Comment:
Missing 'function'. Again, it might work in the latest js but it doesn't on my system.